### PR TITLE
wr-95: Implement avatar upload upon Profile editing

### PR DESCRIPTION
**Pull request by drMangust**
image image
 image

### CODE REVIEW

File: frontend/src/pages/profile/components/avatar-wrapper/avatar-wrapper.tsx [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+          username={`${user.firstName} ${user.lastName}`}
+          avatarUrl={previewUrl}
+          className={styles.avatar}
+        />
+        <input
+          className="visually-hidden"
+          id="avatarId"
+          onChange={handleUploadUserAvatar}
+          type="file"
+          accept=".jpg, .jpeg, .png, .webp, .bmp, .svg"
```
Comment (andanf-e) [inline frontend/src/pages/profile/components/avatar-wrapper/avatar-wrapper.tsx]: There is `SUPPORTED_FILE_TYPES` const on backend side - I believe you can move it to shared and use it also here

File: frontend/src/pages/profile/components/avatar-wrapper/avatar-wrapper.tsx [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+  const handleUploadUserAvatar = useCallback(
+    (event_: ChangeEvent<HTMLInputElement>): void => {
+      const [image] = event_.target.files ?? [];
+
+      if (image) {
+        const formData = new FormData();
+        formData.append('file', image);
+
+        void fileApi
+          .uploadFile(formData)
```
Comment (andanf-e) [inline frontend/src/pages/profile/components/avatar-wrapper/avatar-wrapper.tsx]:
```suggestion
        void fileActions
          .upload(formData)
          .unwrap()
```

File: frontend/src/pages/profile/components/avatar-wrapper/avatar-wrapper.tsx [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+          type="file"
+          accept=".jpg, .jpeg, .png, .webp, .bmp, .svg"
+        />
+      </label>
+      <div className={styles.imageGroupControls}>
+        <Button
+          className={getValidClassNames(styles.control, styles.imageUpdate)}
+          onClick={handleUpdateAvatar}
+          label="Update"
+        />
```
Comment (andanf-e) [inline frontend/src/pages/profile/components/avatar-wrapper/avatar-wrapper.tsx]: I believe we don't need separate button for update - as file input is already an updating field
image
Just add a lighter or darker (what would look better) background with edit icon on hovering field with avatar - to show that it's editable

File: frontend/src/pages/profile/components/avatar-wrapper/avatar-wrapper.tsx [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+        <Button
+          className={getValidClassNames(styles.control, styles.imageRemove)}
+          onClick={handleRemoveAvatar}
+          label="Remove"
+        />
```
Comment (andanf-e) [inline frontend/src/pages/profile/components/avatar-wrapper/avatar-wrapper.tsx]: Please make remove button looking like this
image
And I believe we won't show remove button when there is on avatar uploaded

**Review by andanf-e — CHANGES_REQUESTED** [review]

**Review by NazarKovtun1 — DISMISSED** [review]

File: frontend/src/pages/profile/components/avatar-wrapper/avatar-wrapper.tsx (hunk 1 above)
Comment (drMangust) [reply frontend/src/pages/profile/components/avatar-wrapper/avatar-wrapper.tsx]: done 
image

- [18812a7](https://github.com/BinaryStudioAcademy/bsa-2023-writorium/pull/118/commits/18812a73e3479c9147d33d042afa330f51f30d13)
- [d9396f7](https://github.com/BinaryStudioAcademy/bsa-2023-writorium/pull/118/commits/d9396f77f887b2b51f756f11a903ee3ccbdba14d)

File: frontend/src/pages/profile/components/avatar-wrapper/avatar-wrapper.tsx (hunk 3 above)
Comment (drMangust) [reply frontend/src/pages/profile/components/avatar-wrapper/avatar-wrapper.tsx]: done

File: frontend/src/pages/profile/components/avatar-wrapper/avatar-wrapper.tsx (hunk 4 above)
Comment (drMangust) [reply frontend/src/pages/profile/components/avatar-wrapper/avatar-wrapper.tsx]: done

File: frontend/src/pages/profile/components/avatar-wrapper/avatar-wrapper.tsx (hunk 2 above)
Comment (drMangust) [reply frontend/src/pages/profile/components/avatar-wrapper/avatar-wrapper.tsx]: done
image

- [8e1bece](https://github.com/BinaryStudioAcademy/bsa-2023-writorium/pull/118/commits/8e1bece2c89e366eaf9f5e6e09df7abf44a467a1)

File: backend/src/packages/files/file.model.ts [hunk 5]
```diff
@@ -1,14 +1,37 @@
+import { Model, type RelationMappings } from 'objection';
+
 import {
   AbstractModel,
   DatabaseTableName,
 } from '~/libs/packages/database/database.js';
+import { composeDatabaseRelationPath } from '~/libs/packages/database/libs/helpers/compose-database-relation-path/compose-database-relation-path.helper.js';
```
Comment (artemmatiushenko1) [inline backend/src/packages/files/file.model.ts]: I guess import path should be
`~/libs/packages/database/libs/helpers/helpers.js`

File: frontend/src/pages/profile/components/avatar-wrapper/avatar-wrapper.tsx [hunk 6]
```diff
@@ -0,0 +1,110 @@
+import { type ChangeEvent, type FC } from 'react';
+
+import { Avatar, ErrorMessage, Icon } from '~/libs/components/components.js';
+import { IconButton } from '~/libs/components/icon-button/icon-button.js';
```
Comment (artemmatiushenko1) [inline frontend/src/pages/profile/components/avatar-wrapper/avatar-wrapper.tsx]: `~/libs/components/components.js `

File: frontend/src/pages/profile/components/avatar-wrapper/avatar-wrapper.tsx [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+          >
+            <Icon iconName="edit" className={styles.iconEdit} />
+          </span>
+        </label>
+        <IconButton
+          iconName="crossMark"
+          className={getValidClassNames(
+            styles.iconButtonRemove,
+            styles.iconButton,
+            previewUrl && isHover && styles.iconButtonRemoveActive,
```
Comment (artemmatiushenko1) [inline frontend/src/pages/profile/components/avatar-wrapper/avatar-wrapper.tsx]: Just wondering why can’t we control hover styles using css `:hover` pseudo class?

File: frontend/src/slices/file/action.ts [hunk 8]
```diff
@@ -0,0 +1,17 @@
+import { createAsyncThunk } from '@reduxjs/toolkit';
+import { type FileUploadResponseDto } from 'shared/build/index.js';
```
Comment (artemmatiushenko1) [inline frontend/src/slices/file/action.ts]: Maybe it should be imported from `types` folder?

File: frontend/src/pages/profile/components/avatar-wrapper/avatar-wrapper.tsx (hunk 7 above)
Comment (andanf-e) [reply frontend/src/pages/profile/components/avatar-wrapper/avatar-wrapper.tsx]: Yep, I believe we don't need separate hook to track hovering

Comment (drMangust) [reply frontend/src/pages/profile/components/avatar-wrapper/avatar-wrapper.tsx]: Oh, thanks. that's a leftover from my attempt to implement this without an additional wrapper. I probably saw what worked and didn't think to make it simple. Thanks for noticing.

- [5d09f35](https://github.com/BinaryStudioAcademy/bsa-2023-writorium/pull/118/commits/5d09f3543d937477cab2ce9aca7f40ab32b51eb5)
- [a2d7299](https://github.com/BinaryStudioAcademy/bsa-2023-writorium/pull/118/commits/a2d72992745623127dafbecda5792f23af2acb4f)

File: backend/src/packages/files/file.model.ts (hunk 5 above)
Comment (drMangust) [reply backend/src/packages/files/file.model.ts]: Thank you, such simple mistakes are often overlooked.
I can't find how to configure auto-imports to see re-exports.

- [65654a1](https://github.com/BinaryStudioAcademy/bsa-2023-writorium/pull/118/commits/65654a1286d0d5908f927dd8b03a652952aea9d5)

File: frontend/src/pages/profile/components/avatar-wrapper/avatar-wrapper.tsx (hunk 6 above)
Comment (drMangust) [reply frontend/src/pages/profile/components/avatar-wrapper/avatar-wrapper.tsx]: Thanks for having my back

- [65654a1](https://github.com/BinaryStudioAcademy/bsa-2023-writorium/pull/118/commits/65654a1286d0d5908f927dd8b03a652952aea9d5)

File: frontend/src/slices/file/action.ts (hunk 8 above)
Comment (drMangust) [reply frontend/src/slices/file/action.ts]: Thanks for having my back

- [65654a1](https://github.com/BinaryStudioAcademy/bsa-2023-writorium/pull/118/commits/65654a1286d0d5908f927dd8b03a652952aea9d5)

File: backend/src/packages/files/file.model.ts [hunk 9]
```diff
@@ -1,14 +1,37 @@
...
+import { UserDetailsModel } from '~/packages/users/user-details.model.js';
 
 class FileModel extends AbstractModel {
   public 'url': string;
 
   public static override get tableName(): string {
     return DatabaseTableName.FILES;
   }
+
+  public static get relationMappings(): RelationMappings {
```
Comment (andanf-e) [inline backend/src/packages/files/file.model.ts]: Not sure if this relation for files is necessary

Comment (drMangust) [reply backend/src/packages/files/file.model.ts]: Yes, you're right.
- [a7d9df5](https://github.com/BinaryStudioAcademy/bsa-2023-writorium/pull/118/commits/a7d9df5f326a071b27cb076c3d01f25cd7e26317)

**Review by andanf-e — DISMISSED** [review]

**Review by andanf-e — DISMISSED** [review]

**Review by mhrechkin — DISMISSED** [review]

**Review by NazarKovtun1 — DISMISSED** [review]

**Review by andanf-e — APPROVED** [review]

**Pull request merged**